lock-api/src/test/java/com/palantir/lock/watch/LockWatchStateUpdaterTest.java

```diff
@@ -0,0 +1,234 @@
+    private static final LockWatchEvent OPEN_LOCKS_EVENT = LockWatchOpenLocksEvent
+            .builder(ImmutableSet.of(DESCRIPTOR), UUID.randomUUID())
+            .build(1);
+    private static final LockToken TOKEN = LockToken.of(UUID.randomUUID());
+
+    private final RangeSet<LockDescriptor> watches = TreeRangeSet.create();
+    private final Map<LockDescriptor, LockWatchInfo> lockWatchState = new HashMap<>();
+    private ConcurrentSkipListSet<UUID> unmatchedOpenLocksEvents = new ConcurrentSkipListSet<>();
+    private final LockWatchStateUpdater updater = new LockWatchStateUpdater(watches, lockWatchState,
+            unmatchedOpenLocksEvents);
+
+    @Test
+    public void watchesAreUpdatedOnSingleEventWhenOpenLocksHappenedBefore() {
+        UUID lockWatchId = UUID.randomUUID();
+
+        LockWatchEvent openLocks = LockWatchOpenLocksEvent
+                .builder(ImmutableSet.of(DESCRIPTOR), lockWatchId)
+                .build(1);
+        LockWatchEvent watchCreatedEvent = LockWatchCreatedEvent
+                .builder(LockWatchRequest.of(ImmutableSet.of(TABLE_1, TABLE_3)), lockWatchId)
+                .build(10);
+
+        openLocks.accept(updater);
+        watchCreatedEvent.accept(updater);
+        assertThat(watches.asRanges())
+                .containsExactlyInAnyOrder(TABLE_1.accept(TO_RANGES_VISITOR), TABLE_3.accept(TO_RANGES_VISITOR));
+    }
+
+    @Test
+    public void watchesAreNotUpdatedOnSingleEventWhenOpenLocksHasNotHappenedWithSameId() {
+        LockWatchEvent watchCreatedEvent = LockWatchCreatedEvent
+                .builder(LockWatchRequest.of(ImmutableSet.of(TABLE_1, TABLE_3)), UUID.randomUUID())
+                .build(10);
+
+        OPEN_LOCKS_EVENT.accept(updater);
+        watchCreatedEvent.accept(updater);
+        assertThat(watches.asRanges()).isEmpty();
+    }
+
+    @Test
+    public void watchesAreUpdatedOnMultipleEvents() {
+        UUID first = UUID.randomUUID();
+        UUID second = UUID.randomUUID();
+
+        LockWatchEvent firstOpenLocks = LockWatchOpenLocksEvent
+                .builder(ImmutableSet.of(DESCRIPTOR), first)
+                .build(1);
+        LockWatchEvent secondOpenLocks = LockWatchOpenLocksEvent
+                .builder(ImmutableSet.of(), second)
+                .build(2);
+
+        LockWatchEvent firstEvent = LockWatchCreatedEvent
+                .builder(LockWatchRequest.of(ImmutableSet.of(TABLE_1)), second).build(10);
+        LockWatchEvent secondEvent = LockWatchCreatedEvent
+                .builder(LockWatchRequest.of(ImmutableSet.of(TABLE_3)), first).build(20);
+
+        firstOpenLocks.accept(updater);
+        secondOpenLocks.accept(updater);
+        firstEvent.accept(updater);
+        secondEvent.accept(updater);
+
+        assertThat(watches.asRanges())
+                .containsExactlyInAnyOrder(TABLE_1.accept(TO_RANGES_VISITOR), TABLE_3.accept(TO_RANGES_VISITOR));
+    }
+
+    @Test
+    public void overlappingWatchesAreMergedOnUpdates() {
```

Comment:
not sure how I feel about this test, it feels dependent on the contract of `RangeSet` (which is that it *must* coalesce) rather than anything we're doing in the code

Reply:
That's fair, but it's critical for us that re-registering watches does not blow up our ranges so I think it's fine to have this just to verify the invariant. Also, `RangeSet` is technically `@Beta`...